File: src/main/java/org/team1/nbe1_3_team01/global/auth/login/filter/CustomUsernamePasswordAuthenticationFilter.java
package org.team1.nbe1_3_team01.global.auth.login.filter;

import com.fasterxml.jackson.databind.ObjectMapper;
import jakarta.servlet.ServletException;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import org.springframework.security.authentication.AuthenticationServiceException;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.AuthenticationException;
import org.springframework.security.web.authentication.AbstractAuthenticationProcessingFilter;
import org.springframework.security.web.util.matcher.AntPathRequestMatcher;
import org.springframework.util.StreamUtils;

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.util.Map;

public class CustomUsernamePasswordAuthenticationFilter extends AbstractAuthenticationProcessingFilter {
    private static final String REQUEST_URL = "/api/login";
    private static final String HTTP_METHOD = "POST";
    private static final String CONTENT_TYPE = "application/json";
    private static final String USERNAME_KEY = "username";
    private static final String PASSWORD_KEY = "password";
    // /login + POST 요청에 매칭
    private static final AntPathRequestMatcher PATH_REQUEST_MATCHER
            = new AntPathRequestMatcher(REQUEST_URL, HTTP_METHOD);

    private final ObjectMapper objectMapper;

    public CustomUsernamePasswordAuthenticationFilter(ObjectMapper objectMapper) {
        super(PATH_REQUEST_MATCHER); // 위에서 설정한 "login" + POST로 온 요청을 처리하기 위해 설정
        this.objectMapper = objectMapper;
    }
    @Override
    public Authentication attemptAuthentication(HttpServletRequest request, HttpServletResponse response) throws AuthenticationException, IOException, ServletException {
        if(request.getContentType() == null || !request.getContentType().equals(CONTENT_TYPE)  ) {
            throw new AuthenticationServiceException("ContentType이 " + request.getContentType() + " 가 아닙니다");
        }

        String messageBody = StreamUtils.copyToString(request.getInputStream(), StandardCharsets.UTF_8);
        Map<String, String> usernamePasswordMap = objectMapper.readValue(messageBody, Map.class);
        String username = usernamePasswordMap.get(USERNAME_KEY);
        String password = usernamePasswordMap.get(PASSWORD_KEY);
        //principal 과 credentials 전달
        UsernamePasswordAuthenticationToken usernamePasswordAuthenticationToken = new UsernamePasswordAuthenticationToken(username, password);

        return this.getAuthenticationManager().authenticate(usernamePasswordAuthenticationToken);
    }
}

File: src/main/java/org/team1/nbe1_3_team01/global/auth/login/filter/CustomUsernamePasswordAuthenticationFilter.kt
package org.team1.nbe1_3_team01.global.auth.login.filter

import com.fasterxml.jackson.databind.ObjectMapper
import jakarta.servlet.ServletException
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import org.springframework.security.authentication.AuthenticationServiceException
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken
import org.springframework.security.core.Authentication
import org.springframework.security.core.AuthenticationException
import org.springframework.security.web.authentication.AbstractAuthenticationProcessingFilter
import org.springframework.security.web.util.matcher.AntPathRequestMatcher
import org.springframework.util.StreamUtils
import java.io.IOException
import java.nio.charset.StandardCharsets

class CustomUsernamePasswordAuthenticationFilter(private val objectMapper: ObjectMapper) :
    AbstractAuthenticationProcessingFilter(PATH_REQUEST_MATCHER) {
    @Throws(
        AuthenticationException::class,
        IOException::class,
        ServletException::class
    )
    override fun attemptAuthentication(request: HttpServletRequest, response: HttpServletResponse): Authentication {
        if (request.contentType == null || request.contentType != CONTENT_TYPE) {
            throw AuthenticationServiceException("ContentType이 " + request.contentType + " 가 아닙니다")
        }

        val messageBody = StreamUtils.copyToString(request.inputStream, StandardCharsets.UTF_8)
        val usernamePasswordMap = objectMapper.readValue(
            messageBody,
            MutableMap::class.java
        ) as Map<*, *>
        val username = usernamePasswordMap[USERNAME_KEY]
        val password = usernamePasswordMap[PASSWORD_KEY]
        //principal 과 credentials 전달
        val usernamePasswordAuthenticationToken = UsernamePasswordAuthenticationToken(username, password)

        return authenticationManager.authenticate(usernamePasswordAuthenticationToken)
    }

    companion object {
        private const val REQUEST_URL = "/api/login"
        private const val HTTP_METHOD = "POST"
        private const val CONTENT_TYPE = "application/json"
        private const val USERNAME_KEY = "username"
        private const val PASSWORD_KEY = "password"

        // /login + POST 요청에 매칭
        private val PATH_REQUEST_MATCHER = AntPathRequestMatcher(REQUEST_URL, HTTP_METHOD)
    }
}
